src/libstd/fs.rs

```diff
@@ -0,0 +1,1474 @@
+pub struct WalkDir {
+    cur: Option<ReadDir>,
+    stack: Vec<io::Result<ReadDir>>,
+}
+
+impl Iterator for WalkDir {
+    type Item = io::Result<DirEntry>;
+
+    fn next(&mut self) -> Option<io::Result<DirEntry>> {
+        loop {
+            if let Some(ref mut cur) = self.cur {
+                match cur.next() {
+                    Some(Err(e)) => return Some(Err(e)),
+                    Some(Ok(next)) => {
+                        let path = next.path();
+                        if path.is_dir() {
+                            self.stack.push(read_dir(&*path));
+                        }
+                        return Some(Ok(next))
+                    }
+                    None => {}
+                }
+            }
+            self.cur = None;
+            match self.stack.pop() {
+                Some(Err(e)) => return Some(Err(e)),
+                Some(Ok(next)) => self.cur = Some(next),
+                None => return None,
+            }
+        }
+    }
+}
+
+/// Utility methods for paths.
+pub trait PathExt {
+    /// Get information on the file, directory, etc at this path.
+    ///
+    /// Consult the `fs::stat` documentation for more info.
+    ///
+    /// This call preserves identical runtime/error semantics with `file::stat`.
+    fn file_attr(&self) -> io::Result<FileAttr>;
+
+    /// Boolean value indicator whether the underlying file exists on the local
+    /// filesystem. Returns false in exactly the cases where `fs::stat` fails.
+    fn exists(&self) -> bool;
+
+    /// Whether the underlying implementation (be it a file path, or something
+    /// else) points at a "regular file" on the FS. Will return false for paths
+    /// to non-existent locations or directories or other non-regular files
+    /// (named pipes, etc). Follows links when making this determination.
+    fn is_file(&self) -> bool;
+
+    /// Whether the underlying implementation (be it a file path, or something
+    /// else) is pointing at a directory in the underlying FS. Will return
+    /// false for paths to non-existent locations or if the item is not a
+    /// directory (eg files, named pipes, etc). Follows links when making this
+    /// determination.
+    fn is_dir(&self) -> bool;
+}
+
+impl PathExt for Path {
+    fn file_attr(&self) -> io::Result<FileAttr> { file_attr(self) }
+
+    fn exists(&self) -> bool { file_attr(self).is_ok() }
+
+    fn is_file(&self) -> bool {
+        file_attr(self).map(|s| s.is_file()).unwrap_or(false)
+    }
+    fn is_dir(&self) -> bool {
+        file_attr(self).map(|s| s.is_dir()).unwrap_or(false)
+    }
+}
+
+/// Changes the timestamps for a file's last modification and access time.
+///
+/// The file at the path specified will have its last access time set to
+/// `atime` and its modification time set to `mtime`. The times specified should
+/// be in milliseconds.
+pub fn change_file_times<P: AsPath + ?Sized>(path: &P, atime: u64, mtime: u64)
```

Comment:
> POSIX API allows to both omit either field or set any of those to the current time.

Hm, could you point me for the documentation for this? This is currently implemented using [`utimes`](http://linux.die.net/man/2/utime) which allows you to pass `NULL` for its parameter (which represents both atime/mtime) and if null will set both to the current time. There doesn't seem to be a way to optionally set one time but not the other.

Additionally, the [`SetFileTime`](https://msdn.microsoft.com/en-us/library/windows/desktop/ms724933%28v=vs.85%29.aspx) API we're using for windows doesn't seem to support setting the time to the *current* time but only optionally not setting one of the times specified.

I definitely agree that there's a disconnect between these two APIs, and it may perhaps be better to just not expose this at all.